File: ui/src/utils/device-fragment.test.ts
import { describe, expect, it } from 'vitest';
import { findDeviceFragment, spliceDeviceFragment } from './device-fragment';

// A config that carries the things a naive splice destroys: comments above and
// inside a device, a trailing device after the target, blank lines, and a
// top-level key after the list.
const CONFIG = `# NIAC demo topology
devices:
  # the edge router
  - name: api-router
    type: router
    ips:
      - "10.10.0.1"
    snmp_agent:
      community: "public" # not the default on purpose

  - name: core-switch
    type: switch
    ips:
      - "10.10.0.2"

segments:
  - vlan: 10
`;

describe('findDeviceFragment', () => {
  it('returns the selected device dedented, so the editor shows a standalone document', () => {
    const fragment = findDeviceFragment(CONFIG, 'api-router');
    expect(fragment).not.toBeNull();
    expect(fragment?.text).toBe(
      `name: api-router
type: router
ips:
  - "10.10.0.1"
snmp_agent:
  community: "public" # not the default on purpose
`,
    );
  });

  it('keeps a comment that sits inside the device', () => {
    expect(findDeviceFragment(CONFIG, 'api-router')?.text).toContain(
      '# not the default on purpose',
    );
  });

  it('stops at the next device rather than swallowing it', () => {
    expect(findDeviceFragment(CONFIG, 'api-router')?.text).not.toContain('core-switch');
  });

  it('finds the last device without running into the next top-level key', () => {
    const fragment = findDeviceFragment(CONFIG, 'core-switch');
    expect(fragment?.text).toBe(
      `name: core-switch
type: switch
ips:
  - "10.10.0.2"
`,
    );
  });

  it('returns null for a device the config does not have', () => {
    expect(findDeviceFragment(CONFIG, 'nope')).toBeNull();
  });

  it('returns null when the config has no devices list at all', () => {
    expect(findDeviceFragment('segments:\n  - vlan: 10\n', 'api-router')).toBeNull();
  });

  it('returns null for YAML that does not parse, rather than guessing', () => {
    expect(findDeviceFragment('devices:\n  - name: [unclosed\n', 'api-router')).toBeNull();
  });
});

describe('spliceDeviceFragment', () => {
  it('replaces only the selected device, leaving every other byte identical', () => {
    const fragment = findDeviceFragment(CONFIG, 'api-router');
    if (!fragment) {
      throw new Error('fixture regressed: api-router not found');
    }
    const next = spliceDeviceFragment(CONFIG, fragment, 'name: api-router\ntype: firewall\n');

    expect(next).toContain('# NIAC demo topology');
    expect(next).toContain('  # the edge router');
    expect(next).toContain('type: firewall');
    expect(next).not.toContain('10.10.0.1');
    // Everything after the edited device survives byte for byte.
    expect(next.slice(next.indexOf('  - name: core-switch'))).toBe(
      CONFIG.slice(CONFIG.indexOf('  - name: core-switch')),
    );
  });

  it('re-indents the fragment to the depth the list uses', () => {
    const fragment = findDeviceFragment(CONFIG, 'core-switch');
    if (!fragment) {
      throw new Error('fixture regressed: core-switch not found');
    }
    const next = spliceDeviceFragment(
      CONFIG,
      fragment,
      'name: core-switch\nips:\n  - "10.0.0.9"\n',
    );

    expect(next).toContain('  - name: core-switch\n    ips:\n      - "10.0.0.9"\n');
  });

  it('round-trips unchanged when the fragment is put back as-is', () => {
    for (const name of ['api-router', 'core-switch']) {
      const fragment = findDeviceFragment(CONFIG, name);
      if (!fragment) {
        throw new Error(`fixture regressed: ${name} not found`);
      }
      expect(spliceDeviceFragment(CONFIG, fragment, fragment.text)).toBe(CONFIG);
    }
  });
});

describe('adjacent devices', () => {
  // The daemon writes every config it saves through yaml.Marshal, which puts
  // no blank line between sequence items. Editing a device in such a config
  // used to consume the next device's `- name:` line: the splice deleted that
  // device and grafted its fields onto the edited one.
  const adjacent = `devices:
  - name: dev-a
    type: router
  - name: dev-b
    type: server
`;

  it('stops the fragment at the next device', () => {
    const fragment = findDeviceFragment(adjacent, 'dev-a');

    expect(fragment?.text).toBe('name: dev-a\ntype: router\n');
  });

  it('leaves the following device intact when the edit is spliced back', () => {
    const fragment = findDeviceFragment(adjacent, 'dev-a');
    if (!fragment) throw new Error('expected a fragment for dev-a');

    const result = spliceDeviceFragment(adjacent, fragment, 'name: dev-a\ntype: firewall\n');

    expect(result).toBe(`devices:
  - name: dev-a
    type: firewall
  - name: dev-b
    type: server
`);
  });

  it('still finds the last device, which has no next sibling', () => {
    expect(findDeviceFragment(adjacent, 'dev-b')?.text).toBe('name: dev-b\ntype: server\n');
  });
});

File: internal/config/device_fragment_test.go
package config_test

import (
	"strings"
	"testing"

	"github.com/MustardSeedNetworks/niac-go/internal/config"
)

const fragmentConfig = `# operator's note, must survive an edit
devices:
  - name: api-router
    type: router
    ips: ["10.10.0.1"]

  - name: core-switch
    type: switch
networks:
  - name: lan
`

func splice(t *testing.T, source, device, replacement string) string {
	t.Helper()

	fragment, ok := config.FindDeviceFragment(source, device)
	if !ok {
		t.Fatalf("no fragment for %q", device)
	}

	return config.SpliceDeviceFragment(source, fragment, replacement)
}

func TestSpliceDeviceFragmentPreservesEverythingElse(t *testing.T) {
	next := splice(t, fragmentConfig, "api-router", "name: api-router\ntype: firewall\n")

	for _, want := range []string{
		"# operator's note, must survive an edit",
		"  - name: api-router\n    type: firewall",
		"  - name: core-switch",
		"networks:",
	} {
		if !strings.Contains(next, want) {
			t.Errorf("spliced config lost %q:\n%s", want, next)
		}
	}

	if strings.Contains(next, "10.10.0.1") {
		t.Errorf("edited device kept a replaced field:\n%s", next)
	}
}

// The comment an operator writes inside a device block is the thing a
// re-serialising save silently deletes, which is why the splice exists.
func TestSpliceDeviceFragmentKeepsACommentInsideTheEditedDevice(t *testing.T) {
	next := splice(t, fragmentConfig, "api-router",
		"name: api-router\ntype: router\n# preserved router edit\n")

	if !strings.Contains(next, "    # preserved router edit") {
		t.Errorf("comment inside the edited device was dropped:\n%s", next)
	}
}

func TestSpliceDeviceFragmentIsIdentityForAnUnchangedBlock(t *testing.T) {
	fragment, ok := config.FindDeviceFragment(fragmentConfig, "api-router")
	if !ok {
		t.Fatal("no fragment")
	}

	body := fragmentConfig[fragment.Start:fragment.End]
	dedented := strings.ReplaceAll(strings.TrimPrefix(strings.TrimSuffix(body, "\n"), "  - "), "\n    ", "\n")

	if next := config.SpliceDeviceFragment(fragmentConfig, fragment, dedented); next != fragmentConfig {
		t.Errorf("round trip changed the document:\ngot:\n%s\nwant:\n%s", next, fragmentConfig)
	}
}

// Adjacent items are how the daemon writes every config it saves, since
// yaml.Marshal puts no blank line between sequence items. Splicing the first
// must not swallow the second.
func TestSpliceDeviceFragmentLeavesAnAdjacentDeviceIntact(t *testing.T) {
	adjacent := "devices:\n  - name: dev-a\n    type: switch\n  - name: dev-b\n    type: server\n"

	next := splice(t, adjacent, "dev-a", "name: dev-a\ntype: firewall\n")

	want := "devices:\n  - name: dev-a\n    type: firewall\n  - name: dev-b\n    type: server\n"
	if next != want {
		t.Errorf("got:\n%s\nwant:\n%s", next, want)
	}
}

func TestFindDeviceFragmentHandlesTheLastDeviceAndTrailingComments(t *testing.T) {
	trailing := "devices:\n  - name: only\n    type: switch\n# a trailing note\n"

	next := splice(t, trailing, "only", "name: only\ntype: router\n")

	if !strings.Contains(next, "# a trailing note") {
		t.Errorf("a comment after the last device was swallowed:\n%s", next)
	}

	if !strings.Contains(next, "    type: router") {
		t.Errorf("edit did not land:\n%s", next)
	}
}

func TestFindDeviceFragmentRejectsWhatItCannotLocate(t *testing.T) {
	for name, source := range map[string]string{
		"unparseable":               "devices: [\n",
		"no devices key":            "networks:\n  - name: lan\n",
		"devices is not a sequence": "devices:\n  name: nope\n",
	} {
		if _, ok := config.FindDeviceFragment(source, "api-router"); ok {
			t.Errorf("%s: located a fragment it should not have", name)
		}
	}

	if _, ok := config.FindDeviceFragment(fragmentConfig, "absent"); ok {
		t.Error("located a device that is not in the config")
	}
}
